### DISCUSSION TITLE
FastAPI dependencies and websocket

### DISCUSSION THREAD

**Discussion by gjanezic** [fastapi/fastapi#9894] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, WebSocket, Request, Depends

async def my_context_dependency(request: Request):
    # Do something with request...
    pass

app = FastAPI(dependencies=[Depends(my_context_dependency)])

@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    pass
```

### Description

% websocat ws://127.0.0.1:8000/ws
websocat: WebSocketError: WebSocketError: Received unexpected status code (500 Internal Server Error)

Error:
TypeError: my_context_dependency() missing 1 required positional argument: 'request'

### Operating System

macOS

### FastAPI Version

0.100.0

### Python Version

3.11.4

**Comment by gjanezic** [comment]
`def my_dependency(request: Request | WebSocket):` as suggested in #8503 does not work (anymore?)

`fastapi.exceptions.FastAPIError: Invalid args for response field! Hint: check that starlette.requests.Request | starlette.websockets.WebSocket is a valid Pydantic field type. If you are using a return type annotation that is not a valid Pydantic field (e.g. Union[Response, dict, None]) you can disable generating the response model from the type annotation with the path operation decorator parameter response_model=None. Read more: https://fastapi.tiangolo.com/tutorial/response-model/`

**Comment by monometa** [comment]
I think I asked the same question recently - https://github.com/Intility/fastapi-azure-auth/issues/155. Perhaps someone can look at it?

**Comment by YuriiMotov** [comment, marked as answer]
FastAPI relies on type annotations to inject instances of `Response` or `WebSocket`.
Since `WebSocket` is not a subclass of `Response` and there is no `Request` in websocket paths, it fails to inject it.

To fix this you can change the type annotation to `HTTPConnection` that is a common ancestor of `Response` and `WebSocket`:

```python
async def my_context_dependency(request: HTTPConnection):
    if isinstance(request, Request):
        # Do something with request...
        pass

```
